jit/jit-insn.c

```diff
@@ -6262,57 +6271,108 @@ jit_insn_return_reg(jit_function_t func, jit_value_t value, int reg)
 }
 
 /*@
- * @deftypefun int jit_insn_setup_for_nested (jit_function_t @var{func}, int @var{nested_level}, int @var{reg})
- * Output an instruction to set up for a nested function call.
- * The @var{nested_level} value will be -1 to call a child, zero to call a
- * sibling of @var{func}, 1 to call a sibling of the parent, 2 to call
- * a sibling of the grandparent, etc.  If @var{reg} is not -1, then
- * it indicates the register to receive the parent frame information.
- * If @var{reg} is -1, then the frame information will be pushed on the stack.
- *
- * You normally wouldn't call this yourself - it is used internally by the
- * CPU back ends to set up the parameters for a nested subroutine call.
+ * @deftypefun int jit_insn_flush_struct (jit_function_t @var{func}, jit_value_t @var{value})
+ * Flush a small structure return value out of registers and back
+ * into the local variable frame.  You normally wouldn't call this
+ * yourself - it is used internally by the CPU back ends to handle
+ * structure returns from functions.
  * @end deftypefun
 @*/
 int
-jit_insn_setup_for_nested(jit_function_t func, int nested_level, int reg)
+jit_insn_flush_struct(jit_function_t func, jit_value_t value)
 {
-	jit_value_t reg_value = jit_value_create_nint_constant(func, jit_type_int, reg);
-	if(!reg_value)
+	if(value)
 	{
-		return 0;
+		jit_value_set_addressable(value);
 	}
+	return create_unary_note(func, JIT_OP_FLUSH_SMALL_STRUCT, value);
+}
+
+/*@
+ * @deftypefun jit_value_t jit_insn_get_frame_pointer (jit_function_t @var{func})
+ * Retrieve the frame pointer of function @var{func}
+ * Returns NULL if out of memory.
+ * @end deftypefun
+@*/
+jit_value_t
+jit_insn_get_frame_pointer(jit_function_t func)
+{
+	return create_dest_note(func, JIT_OP_RETRIEVE_FRAME_POINTER,
+		jit_type_void_ptr);
+}
+
+static jit_value_t
+find_frame_of(jit_function_t func, jit_function_t target,
+	jit_function_t func_start, jit_value_t frame_start)
+{
+	/* Find the nesting level */
+	int nesting_level = 0;
+	jit_function_t current_func = func_start;
+	while(current_func != 0 && current_func != target)
+	{
+		/* Ensure that current_func has a function builder */
+		if(!_jit_function_ensure_builder(current_func))
+		{
+			return 0;
+		}
+
+		if(!current_func->parent_frame)
+		{
+			/* One of the ancestors is not correctly set up */
+			return 0;
+		}
 
-	if(nested_level < 0)
+		current_func = current_func->nested_parent;
+		nesting_level++;
+	}
+	if(!current_func)
 	{
-		return create_unary_note(func, JIT_OP_SETUP_FOR_NESTED, reg_value);
+		/* The value is not accessible from this scope */
+		return 0;
 	}
 
-	jit_value_t nested_level_value =
-		jit_value_create_nint_constant(func, jit_type_int, nested_level);
-	if(!nested_level_value)
+	current_func = func_start;
+	while(frame_start != 0 && nesting_level-- > 0)
+	{
+		jit_value_ref(func, current_func->parent_frame);
+		_jit_gen_fix_value(current_func->parent_frame);
```

Comment:
_jit_gen_* functions are normally called during code generation so it is inappropriate to call one here. But I think for nested functions we could reserve a position for the parent pointer just at the beginning of the stack frame.

Reply:
`_jit_gen_fix_value` is used later in line [6427](https://github.com/ademakov/libjit/pull/15/files/5281baec4507da565982f1f9335f85a3c3af4461#diff-f63607a44dc9f6865c40785657887e17R6427) and [6444](https://github.com/ademakov/libjit/pull/15/files/5281baec4507da565982f1f9335f85a3c3af4461#diff-f63607a44dc9f6865c40785657887e17R6444) again. It is used on every value imported by a nested function.

On one hand this means that reserving stack space isn't really an option as we don't know how many values we are importing and thus don't know how much space we need to reserve.
On the other hand reserving frame space is exactly what `_jit_gen_fix_value` does. It increments the blocks frame size and assigns a frame offset to the value. Apart from alignment rules it is the same on all backends.

I agree that for good coupling the frontend should not make calls to `jit_gen_*` functions. We could create a frontend function for managing stack space, but it would need knowledge about stack alignment etc.. Alternatively we could re-add an IMPORT opcode which calls `_jit_gen_fix_value` and then loads the value.

What do you think?

Reply:
It looks okay to re-add IMPORT opcode although make it simpler than it was before. It will take frame_start value you find here and the value it needs to import.